hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/HederaLifecyclesImpl.java

```diff
@@ -96,15 +111,155 @@ public void onUpdateWeight(
                     configAddressBook.updateWeight(id, stakingInfo.weight());
                     configNodeIds.remove(id);
                 } else {
-                    logger.info("Node {} is deleted from configAddressBook during upgrade ", id);
+                    // We need to validate and mark any node that are removed during upgrade as deleted
+                    // and also set the weight to 0
+                    stakingInfoState.put(
+                            EntityNumber.newBuilder().number(id.id()).build(),
+                            stakingInfo.copyBuilder().deleted(true).weight(0).build());
+                    logger.info(
+                            "Node {} is deleted from configAddressBook during upgrade "
+                                    + "and is marked deleted in state",
+                            id);
                 }
             }
             // for any newly added nodes that doesn't exist in state, weight should be set to 0
             // irrespective of the weight provided in config.txt
-            configNodeIds.forEach(nodeId -> configAddressBook.updateWeight(nodeId, 0));
+            if (!configNodeIds.isEmpty()) {
+                configNodeIds.forEach(nodeId -> {
+                    configAddressBook.updateWeight(nodeId, 0);
+                    logger.info(
+                            "Node {} is newly added in configAddressBook during upgrade "
+                                    + "with weight 0 in configAddressBook",
+                            nodeId);
+                });
+                // update the state with new nodes and set weight to 0
+                addAdditionalNodesToState(
+                        stakingInfoState, configNodeIds, context.getConfiguration(), configAddressBookSize);
+            }
+
+            if (stakingInfoState.isModified()) {
+                ((WritableKVStateBase) stakingInfoState).commit();
+            }
         } else {
-            logger.warn("Token service state is empty to update weights from StakingInfo Map");
+            // When doing the first upgrade from 0.48 to 0.49, we will call updateWeight before BBM migration.
```

Comment:
yes because it will be `MerkleHederaState` with `TokenService`. In 0.48 to 0.49 , if we are calling `updateWeight` after `migrate`, it will still be non-empty. But, platform is calling `updateWeight` before `migrate`

Reply:
Hmmm, sorry for not being clear. According to the comment, this else-branch should only be executed when we upgrade from 0.48 to 0.49. I wonder if the condition (`tokenServiceState.isEmpty()`) is sufficient to isolate this case. What happens during Genesis, for example? Is there any other case possible where `tokenServiceState.isEmpty()` is `true`?